### PR TITLE
RFC: #157 Allow other packages to register client capabilities.

### PR DESCRIPTION
**Pull request by tigersoldier**
Adds two functions for registering/unregistering client capabilities:
`lsp-register-client-capabilities` and `lsp-unregister-client-capabilities`.

Registered capabilities are stored in the `lsp--extra-client-capabilities`
alist in the format of `(PACKAGE-NAME . CAPS)` cons cell. CAPS can be either
a plist, nil, or a function.

All registerd client capabilities are merged into the default client
capabilities. The `lsp--merge-two-plists` function implements a deep merging
algorithm for plists. Existing leaf capability entry cannot be overwritten.

### CODE REVIEW

**Comment by vibhavp** [comment]
How would this be used by `company-lsp`? AFAICT, `lsp-register-client-capabilities` would have to be called before the document/workspace is initialized, right?

**Comment by tigersoldier** [comment]
My current solution is to call it when `company-lsp` is required. See tigersoldier/company-lsp@be7fa992a0d3f2d42f5081a144164693d0e93e21

The idea of accepting a function is to allow packages to detect whether they are used an not, and return nil if they are not used. I haven't tested if the order of major mode hooks can affect the results. Ideally `initialize` request is delayed until all the major mode hooks are run so that all the variables are ready.

**Comment by bkchr** [comment]
Couldn't we introduce a hook that is called, before a workspace is initialized? This hook then could be used to register further client capabilities.

**Comment by vibhavp** [comment]
@bkchr Added in ed0c4523665c44d8a836d854dbdd60d9ac19b416

@tigersoldier Instead of associating with package names, why not use a documentation string instead?

**Comment by bkchr** [comment]
@vibhavp ty for the fast implementation :)

What do you mean by documentation string? Just an arbitrary string like "company-lsp-caps"?

**Comment by tigersoldier** [comment]
@vibhavp @bkchr before initialization hook is great. Let me update the implementation to limit the registration scope in workspace.

Why strings are more preferred than symbols?

**Comment by tigersoldier** [comment]
Update the PR to register client capabilities into workspaces and run the pre initialize hook before getting client capabilities.

File: lsp-methods.el [hunk 1]
```diff
@@ -344,6 +409,48 @@ disappearing, unset all the variables related to it."
   "Client Text document capabilities according to LSP."
   `(:synchronization (:didSave t)))
 
+;;;###autoload
```
Comment (bkchr) [inline lsp-methods.el]: If this function is called in the workspace hook, the autoload can be removed? As lsp-mode should already be loaded, when the hook is executed.

**Comment by bkchr** [comment]
Looks good so far, maybe next time use `git rebase` to get the changes from upstream master. But that is only a personal preference :)

**Comment by tigersoldier** [comment]
@bkchr removed autoload

I use git rebase on my own projects, but I am not familiar with the pull request system of github so not sure what the best practice is. Rebased the PR and squashed the commits.

**Comment by bkchr** [comment]
If you rebase and force push, github automatically filters out all commits that are already in the branch you are want to merge against :) 
But squashing is also nice. :)

@vibhavp any further concerns?

**Comment by vibhavp** [comment]
Yeah, looks good. Thanks!

**Pull request merged**